src/zkevm_specs/evm/execution/end_block.py

```diff
@@ -4,69 +4,102 @@
+#
+# We prove (C) by showing that the transaction ID after the last one processed
+# corresponds to a padding tx in the TxTable, which enforces that once a
+# padding txs appears in the table, the rest will also be padding.  This way,
+# we know that if a padding tx with tx_id exist, then at most there are tx_id-1
+# non-padding txs.  In case we have exhausted the TxTable with txs, we won't
+# have any padding txs; so we skip the lookup.
 #
-# We prove (4) in the circuit implementation by constraining that the last
+# We prove (D) in the circuit implementation by constraining that the last
 # execution step at the end of the EVM circuit is an EndBlock.  When the number
 # of steps is less than the EVM circuit height, we pad at the end with
 # EndBlock.  This will require the EndBlock to have height = 1 in the circuit,
 # which can be achieved after reducing the number of cells used in the state
 # selector.
 
 # Count the max number of txs that the TxTable can hold by counting rows of
-# fixed fields + padding rows in the fixed fields section.
+# type CallerAddress.
 def get_tx_table_max_txs(table: Set[TxTableRow]) -> int:
-    fixed_field_count = 0
-    for row in table:
-        if (row.field_tag != TxContextFieldTag.CallData) or (
-            row.field_tag == TxContextFieldTag.Pad and row.tx_id != 0
-        ):
-            fixed_field_count += 1
-    return fixed_field_count // TxContextFieldTag.TxSignHash
+    return len([row for row in table if row.field_tag == TxContextFieldTag.CallerAddress])
 
 
 def end_block(instruction: Instruction):
+    max_txs = get_tx_table_max_txs(instruction.tables.tx_table)
+    max_rws = len(instruction.tables.rw_table)
+    total_txs = FQ(
+        len(
+            [
+                tx_row
+                for tx_row in instruction.tables.tx_table
+                if tx_row.field_tag == TxContextFieldTag.CallerAddress
+                and tx_row.value.expr() != FQ(0)
+            ]
+        )
+    )
+    last_rw_counter = instruction.curr.rw_counter - 1
+
+    total_rws = last_rw_counter
+    if total_txs.n > 0:
+        # If the block is not empty, we will do a call_context lookup
+        total_rws = last_rw_counter + 1
+
     if instruction.is_last_step:
-        # 1. Verify rw_counter counts to the same number of meaningful rows in
+        # 1. Constraint total_rws and total_txs witness values depending on the empty block case.
+        if total_rws == 0:
+            # 1a. total_txs is 0 in empty block
+            instruction.constrain_equal(total_txs, FQ(0))
+        else:
+            # 1b. total_txs matches the tx_id that corresponds to the final step.
+            instruction.constrain_equal(
+                instruction.call_context_lookup(CallContextFieldTag.TxId), total_txs
+            )
+            # total_rws is last_rw_counter + 1.  extra 1 from tx_id lookup.
+            instruction.constrain_equal(total_rws, last_rw_counter + 1)
```

Comment:
With `total_rws` defined as my proposal above, this check seems unnecessary.

Reply:
Updated in https://github.com/privacy-scaling-explorations/zkevm-specs/pull/286/commits/21e6a522ce2e0fbcdfb78759c4526d108fbf94ba